src/liballoc/task.rs

```diff
@@ -0,0 +1,132 @@
+    fn wake(arc_self: &Arc<Self>);
+
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed. This function is like `wake`, but can only be called from the
+    /// thread on which this `Wake` was created.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    #[inline]
+    unsafe fn wake_local(arc_self: &Arc<Self>) {
+        Self::wake(arc_self);
+    }
+}
+
+struct ArcWrapped<T>(PhantomData<T>);
+
+unsafe impl<T: Wake + 'static> UnsafeWake for ArcWrapped<T> {
+    #[inline]
+    unsafe fn clone_raw(&self) -> Waker {
+        let me: *const ArcWrapped<T> = self;
+        let arc = (*(&me as *const *const ArcWrapped<T> as *const Arc<T>)).clone();
+        Waker::from(arc)
+    }
+
+    #[inline]
+    unsafe fn drop_raw(&self) {
+        let mut me: *const ArcWrapped<T> = self;
+        let me = &mut me as *mut *const ArcWrapped<T> as *mut Arc<T>;
+        ptr::drop_in_place(me);
+    }
+
+    #[inline]
+    unsafe fn wake(&self) {
+        let me: *const ArcWrapped<T> = self;
+        T::wake(&*(&me as *const *const ArcWrapped<T> as *const Arc<T>))
+    }
+
+    #[inline]
+    unsafe fn wake_local(&self) {
+        let me: *const ArcWrapped<T> = self;
+        T::wake_local(&*(&me as *const *const ArcWrapped<T> as *const Arc<T>))
+    }
+}
+
+impl<T> From<Arc<T>> for Waker
+    where T: Wake + 'static,
+{
+    fn from(rc: Arc<T>) -> Self {
+        unsafe {
+            let ptr = mem::transmute::<Arc<T>, NonNull<ArcWrapped<T>>>(rc);
+            Waker::new(ptr)
+        }
+    }
+}
+
+/// Creates a `LocalWaker` from a local `wake`.
+///
+/// This function requires that `wake` is "local" (created on the current thread).
+/// The resulting `LocalWaker` will call `wake.wake_local()` when awoken, and
+/// will call `wake.wake()` if awoken after being converted to a `Waker`.
+#[inline]
+pub unsafe fn local_waker<W: Wake + 'static>(wake: Arc<W>) -> LocalWaker {
+    let ptr = mem::transmute::<Arc<W>, NonNull<ArcWrapped<W>>>(wake);
+    LocalWaker::new(ptr)
+}
+
+struct NonLocalAsLocal<T>(ArcWrapped<T>);
+
+unsafe impl<T: Wake + 'static> UnsafeWake for NonLocalAsLocal<T> {
+    #[inline]
+    unsafe fn clone_raw(&self) -> Waker {
+        self.0.clone_raw()
+    }
+
+    #[inline]
+    unsafe fn drop_raw(&self) {
+        self.0.drop_raw()
+    }
+
+    #[inline]
+    unsafe fn wake(&self) {
+        self.0.wake()
+    }
+
+    #[inline]
+    unsafe fn wake_local(&self) {
+        // Since we're nonlocal, we can't call wake_local
+        self.0.wake()
+    }
+}
+
+/// Creates a `LocalWaker` from a non-local `wake`.
+///
+/// This function is similar to `local_waker`, but does not require that `wake`
+/// is local to the current thread. The resulting `LocalWaker` will call
+/// `wake.wake()` when awoken.
+#[inline]
+pub fn local_waker_from_nonlocal<W: Wake + 'static>(wake: Arc<W>) -> LocalWaker {
```

Comment:
Could this be a `From` coercion? Or do you prefer to keep it explicit?

The fact that we can't add it as an inherent constructor is a shame. But we can iterate on this later.

Reply:
`LocalWaker` isn't defined in `liballoc`, so we can't add a `From<impl Wake + 'static> for LocalWaker` here. In any case, I think it might be a bit too tricky for users to figure out that `wake_local` won't be called via the `From` impl.

Reply:
:+1: